coral-incremental/src/main/java/com/linkedin/coral/incremental/RelNodeGenerationTransformer.java

```diff
@@ -0,0 +1,566 @@
+   * <p>
+   * will be transformed to:
+   * <p>
+   * SELECT * FROM test.bar1_prev JOIN test.bar2_prev ON test.bar1_prev.x = test.bar2_prev.x
+   */
+  public RelNode convertRelPrev(RelNode originalNode) {
+    RelShuttle converter = new RelShuttleImpl() {
+      @Override
+      public RelNode visit(TableScan scan) {
+        RelOptTable originalTable = scan.getTable();
+        List<String> incrementalNames = new ArrayList<>(originalTable.getQualifiedName());
+        String deltaTableName = incrementalNames.remove(incrementalNames.size() - 1) + PREV_SUFFIX;
+        incrementalNames.add(deltaTableName);
+        RelOptTable incrementalTable =
+            RelOptTableImpl.create(originalTable.getRelOptSchema(), originalTable.getRowType(), incrementalNames, null);
+        return LogicalTableScan.create(scan.getCluster(), incrementalTable);
+      }
+
+      @Override
+      public RelNode visit(LogicalJoin join) {
+        RelNode left = join.getLeft();
+        RelNode right = join.getRight();
+        RelNode prevLeft = convertRelPrev(left);
+        RelNode prevRight = convertRelPrev(right);
+        RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+
+        LogicalProject p3 = createProjectOverJoin(join, prevLeft, prevRight, rexBuilder);
+
+        return p3;
+      }
+
+      @Override
+      public RelNode visit(LogicalFilter filter) {
+        RelNode transformedChild = convertRelPrev(filter.getInput());
+
+        return LogicalFilter.create(transformedChild, filter.getCondition());
+      }
+
+      @Override
+      public RelNode visit(LogicalProject project) {
+        RelNode transformedChild = convertRelPrev(project.getInput());
+        return LogicalProject.create(transformedChild, project.getProjects(), project.getRowType());
+      }
+
+      @Override
+      public RelNode visit(LogicalUnion union) {
+        List<RelNode> children = union.getInputs();
+        List<RelNode> transformedChildren =
+            children.stream().map(child -> convertRelPrev(child)).collect(Collectors.toList());
+        return LogicalUnion.create(transformedChildren, union.all);
+      }
+
+      @Override
+      public RelNode visit(LogicalAggregate aggregate) {
+        RelNode transformedChild = convertRelPrev(aggregate.getInput());
+        return LogicalAggregate.create(transformedChild, aggregate.getGroupSet(), aggregate.getGroupSets(),
+            aggregate.getAggCallList());
+      }
+    };
+    return originalNode.accept(converter);
+  }
+
+  /**
+   * Transforms the given relational algebra tree to ensure a uniform format by recursively processing its nodes.
+   * This transformation involves:
+   * - For LogicalJoin nodes: recursively processing their children, and optionally creating a projection over the join
+   *   if the join is in the needsProj set. (when the Join don't have a LogicalProject as its parent)
+   * - For other RelNodes: recursively processing their inputs to ensure uniformity.
+   * <p>
```

Comment:
Not clear what this achieves.

Reply:
Does the example below show what it's doing?

Reply:
I am referring to "to ensure uniformity". The idea about the whole method is to ensure uniformity. Making it exclusive to "other RelNodes" does not make sense.